### ISSUE TITLE
Backward compatibility tests for extensions

### ISSUE THREAD

**Issue by lubennikovaav** [neondatabase/neon#9206]
Add backward compatibility tests which we could use when we want to bump version from N to M
part of  https://github.com/neondatabase/company_projects/issues/291

## Motivation

Currently, we update extensions by changing version in the Dockerfile.compute-node.
It is hard to tell, if new extension binary works with objects created by previous version.

## DoD

We have test coverage that allows us to perform extension updates safely.

## Implementation ideas

1. First idea is to add `CREATE EXTENSION` for all available extensions to backward_compatibility [snapshot](https://github.com/neondatabase/neon/blob/main/test_runner/regress/test_compatibility.py#L117). 
This will help to catch simple cases, when extension is missing in the new image.

2. Testing backward compatibility is a very broad task, because extension may create lots of SQL objects (functions, tables, views) and change their signatures and internal logic arbitrary. One thing that is relatively easy to test is that new `.so` has all the functions, needed for previously installed version of extension.  We can compare function signatures using query like this and comparing results before and after image update:
```
SELECT e.extname, p.proname, p.proargtypes, p.prorettype
FROM pg_catalog.pg_extension AS e
    INNER JOIN pg_catalog.pg_depend AS d ON (d.refobjid = e.oid)
    INNER JOIN pg_catalog.pg_proc AS p ON (p.oid = d.objid)
WHERE d.deptype = 'e'
ORDER BY e.extname, p.proname;
```

3. We can also run `ALTER EXTENSION extname UPDATE` on a backward compatibility snapshot data, to ensure that update path is present.
This probably won't work for some extensions, test `DROP extension + CREATE extension` for them.

## Other related tasks and Epics

need review from @neondatabase/developer-productivity 
Ideas are welcome.

**Comment by Bodobolero** [comment]
We need to distinguish between the extension version (version of the signature) and the extension binary version.
A full test coverage would require:
- creating a project with old version of Neon compute image with old version of extension
- testing the "old project" with old version of extension signature with the new extension binary in the new compute image: testing here means run the extension specific test scripts and verify correctness of results (see https://neondb.slack.com/archives/C0732L0A4AH/p1715937229149149?thread_ts=1715930923.849179&cid=C0732L0A4AH as an example where the signature was still supported by the new binary but the semantics slightly changed)
- upgrading the extension version while using the new compute image with the new compute image (to test that the upgrade works correctly)
- and testing it again (to test that the signatures and semantics still are correct after the upgrade

In summary a full test coverage would mean that we test the full matrix of NxMxOxPx2
- N postgres versions supported by Neon
- M extensions supported by Neon 
- O extension binary version (compute image version) and extension version (signature version) used when creating the extension
- P1 extension binary version (compute image version) that we are testing (after upgrading binary version) while maintaining original signature version
- P2 extension binary version(compute image version) that we are testing after upgrading binary version and after upgrading the extension signature to latest version

I think practically we should limit it to a subset.

Suggestions on subset:

- define metrics that allows us to measure which extensions are created in which projects
- N: only test last 2 postgres versions (newest and the one prior)
- M: only test extensions used in projects with more than 10 paying customers
- O: only test oldest binary version and upgrade to newest binary version (not intermediate ones)
- P1: only test newest compute image version
- P2: only test newest extension signature version

**Comment by lubennikovaav** [comment]
>testing the "old project" with old version of extension signature with the new extension binary in the new compute image

This is exactly what I suggested above. We already have `backward_compatibility` snapshots infrastracture.
We can and save a snapshot of postgres with old version of extension installed, and run new compute (+extension) binary on this snapshot.

Alternatively, we can use cloud setup and use branches as data snapshots.
I think this is better idea, but I don't know how we can test PRs against real cloud neon and use compute image from PR there.
We need to create "compute preview env"  like: take latest branch a test argument --> branch it ---> pin image for endpoint --> spin the test, save the branch as base for next runs).
WDYT?

>means run the extension specific test scripts and verify correctness of results

I suggest to start with something simpler (function signature check).

With extension specific tests, we again rely on third-party extension maintainer to provide good tests.
If they, for example, removed a function between releases and also removed it from tests, the tests will pass.
Or should we also use test set from previous release? This seems quite complicated to me.

**Comment by bayandin** [comment]
Given that we currently don't build extensions (for regression tests), I'd consider writing cloud e2e test (i.e. not cloud Neon) for extensions compatibility tests.
We can get the previous version of compute (from GitHub releases/git tags, for example) to do some preparation, and then start new version of compute (or version from a PR) to perform checks themself

**Comment by lubennikovaav** [comment]
I don't think this test belongs to cloud repo, or at least I don't understand how to untangle the dependencies.

Simple scenario, even without backward compatibility checks:
I have neon PR that adds new extension (or bumps version of existing one)
I want to add a test that checks that extension is installed with expected version

If the test lives in cloud e2e, I have to add 2 commits to 2 different repos atomically (because they depend on each other. test will fail while extension is not present in compute image, and I cannot run the test, until it is committed to cloud repo)

That's why I want to add neon cloud e2e, similar to cloud_regress
Am I missing something?

**Comment by a-masterov** [comment]
We are already building the image for testing extensions. We may want to use it with docker-composer:

1. Start the docker-composer environment with the old compute image
2. Create extensions on the old version
3. Shutdown the old version of the compute image and start the new one
4. Run the tests
